otwcode/otwarchive
# AO3-5814 Allow skipping Codeship tests using commit messages

## Issue

https://otwarchive.atlassian.net/browse/AO3-5814

## Purpose

Update Codeship scripts to skip the setup and test pipelines when the commit message contains "skip codeship tests" (with square brackets).

Other updates:

- Skip installing Redis 3.2.1, as Codeship's version is now later (4.0.9).
- Skip installing MySQL 5.7.22, as Codeship's version is now later (5.7.27). The default instance uses port 3306.
- Use Codeship's script for installing and starting Elasticsearch.
- Remove code for resetting coverage in the retry script, as we no longer generate coverage on Codeship.
- Generate just the test database using `rake db:create` instead of `rake db:create:all`.
- Split up test_prepare.sh. This script is run twice (once for specs, once for features) but most of it only needs to be run once (e.g. installing dependencies).
- Download wkhtmltopdf into a directory cached between builds.

## Testing Instructions

Codeship builds should pass.

Before this is merged, we need to update project settings on Codeship.

New "Setup Commands":

```bash
rvm use $(cat .ruby-version) --install
./script/codeship/setup.sh
```

New "Configure Test Pipelines":

```bash
# The expectation is that this script is copied in to the test area
# on Codeship rather than run as a single script as having timings for
# each stage is useful.
#
# Adding "|| :" to the end of a command will force it to succeed,
# useful for ignoring failed tests
export RAILS_ENV=test
export CUCUMBER_FORMAT=Ao3Cucumber::Formatter

./script/codeship/test_prepare.sh
./script/codeship/try.sh rspec "bundle exec rspec spec/controllers"
./script/codeship/try.sh rspec "bundle exec rspec spec/miscellaneous"
./script/codeship/try.sh rspec "bundle exec rspec spec/models"

./script/codeship/test_prepare.sh
./script/codeship/try.sh admins             "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh bookmarks          "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh collections        "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh comments_and_kudos "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh gift_exchanges     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh importing          "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh other_a            "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh other_b            "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_a     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_b     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh prompt_memes_c     "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh search             "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh tag_sets           "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh tags_and_wrangling "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh users              "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
./script/codeship/try.sh works              "bundle exec cucumber --tags ~@browserstack features/\$TEST_RUN"
```
